enterprise/server/registry/registry_service/registry_service.go

```diff
@@ -0,0 +1,980 @@
+			return nil, status.PermissionDeniedErrorf("could not retrieve image manifest: %s", err)
+		}
+		return nil, status.UnavailableErrorf("could not retrieve manifest from remote: %s", err)
+	}
+
+	remoteImg, err := r.targetImageFromDescriptor(remoteDesc, req.GetPlatform())
+	if err != nil {
+		return nil, err
+	}
+
+	// Check whether the supplied credentials are sufficient to access the
+	// remote image.
+	if err := r.checkAccess(ctx, imageRef, remoteImg, authenticator); err != nil {
+		return nil, err
+	}
+
+	// If we got here then it means the credentials are valid for the remote
+	// repo. Now we can return the optimized image ref to the client.
+
+	manifest, err := r.getCachedManifest(ctx, remoteDesc.Digest.String())
+	if err != nil {
+		return nil, status.UnavailableErrorf("could not check for cached manifest %s: %s", imageRef, err)
+	}
+	if manifest != nil {
+		log.CtxInfof(ctx, "Using cached manifest information")
+	} else {
+		convertedManifest, err := r.convert(ctx, remoteDesc)
+		if err != nil {
+			return nil, status.UnknownErrorf("could not convert image: %s", err)
+		}
+		manifest = convertedManifest
+	}
+
+	encodedImageName := strings.ToLower(imageNameEncoding.EncodeToString([]byte(imageRef.Context().Name())))
+
+	return &regpb.GetOptimizedImageResponse{
+		OptimizedImage: fmt.Sprintf("%s@%s", encodedImageName, manifest.Digest),
+	}, nil
+}
+
+func (r *registry) handleRegistryRequest(w http.ResponseWriter, req *http.Request) {
+	ctx := req.Context()
+	log.CtxInfof(ctx, "%s %q", req.Method, req.RequestURI)
+	// Clients issue a GET /v2/ request to verify that this is a registry
+	// endpoint.
+	if req.RequestURI == "/v2/" {
+		w.WriteHeader(http.StatusOK)
+		return
+	}
+	// Request for a manifest or image index.
+	if m := manifestReqRE.FindStringSubmatch(req.RequestURI); len(m) == 3 {
+		r.handleManifestRequest(w, req, m[1], m[2])
+		return
+	}
+	// Request for a blob (full layer or layer chunk).
+	if m := blobReqRE.FindStringSubmatch(req.RequestURI); len(m) == 3 {
+		r.handleBlobRequest(w, req, m[1], m[2])
+		return
+	}
+	http.NotFound(w, req)
+}
+
+func (r *registry) Start(env environment.Env) error {
+	mux := http.NewServeMux()
+
+	handler := http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		r.handleRegistryRequest(w, req)
+	})
+	mux.Handle("/v2/", filters.RequestID(handler))
+	mux.Handle("/readyz", env.GetHealthChecker().ReadinessHandler())
+	mux.Handle("/healthz", env.GetHealthChecker().LivenessHandler())
+
+	regRegistryServices := func(server *grpc.Server, env environment.Env) {
+		regpb.RegisterRegistryServer(server, r)
+	}
+
+	if err := grpc_server.RegisterGRPCServer(env, *gRPCPort, regRegistryServices); err != nil {
+		return status.UnknownErrorf("Could not setup GRPC server: %s", err)
+	}
+	if err := grpc_server.RegisterGRPCSServer(env, *gRPCSPort, regRegistryServices); err != nil {
+		return status.UnknownErrorf("Could not setup GRPCS server: %s", err)
+	}
```

Comment:
Can we run this on 5151 (the existing internal grpc service used by ddisk?) and then add a new port for internal HTTP traffic?

Reply:
https://github.com/buildbuddy-io/buildbuddy-internal/blob/master/enterprise/deployment/dev/us-west1-b/buildbuddy-app.dev.yaml#L13

Reply:
I've added new "standard" ports for internal gRPC & http services.
Migrating distributed cache over to the standard port is more involved. For one, we still need to support the distributed_cache.listen_addr flag because onprem customers may already be using it and it's not a just a simple port flag. We could additionally serve the distributed cache RPCs on the new ports and migrate over, but I don't see a simple way of getting rid of the separate serving code for existing users.